src/input/process.rs

```diff
@@ -80,6 +83,46 @@ pub fn read_processes(
     )
 }
 
+/// Perform consistency checks for commodity flows.
+fn validate_commodities(
+    commodities: &HashMap<Rc<str>, Rc<Commodity>>,
+    flows: &HashMap<Rc<str>, Vec<ProcessFlow>>,
+) -> Result<()> {
+    for (commodity_id, commodity) in commodities {
+        if commodity.kind == CommodityType::SupplyEqualsDemand {
+            validate_sed_commodity(commodity_id, commodity, flows)?;
+        }
+    }
+    Ok(())
+}
+
+fn validate_sed_commodity(
+    commodity_id: &Rc<str>,
+    commodity: &Rc<Commodity>,
+    flows: &HashMap<Rc<str>, Vec<ProcessFlow>>,
+) -> Result<()> {
+    let mut has_producer = false;
+    let mut has_consumer = false;
+
+    for flow in flows.values().flatten() {
+        if Rc::ptr_eq(&flow.commodity, commodity) {
+            if flow.flow > 0.0 {
+                has_producer = true;
+            } else if flow.flow < 0.0 {
+                has_consumer = true;
+            }
+        }
+    }
+
+    ensure!(
+        has_producer && has_consumer,
+        "Commodity {} of 'SED' type must have both producer and consumer processes",
+        commodity_id
+    );
+
+    Ok(())
```

Comment:
You can return early here if you've already found a producer and consumer:

```suggestion
            if flow.flow > 0.0 {
                has_producer = true;
            } else if flow.flow < 0.0 {
                has_consumer = true;
            }

            if has_producer && has_consumer {
                return Ok(());
            }
        }
    }

    anyhow!(
        "Commodity {} of 'SED' type must have both producer and consumer processes",
        commodity_id
    )
```

NB: You'll also need to `use anyhow::anyhow` for this to work.